hathor/indexes/rocksdb_timestamp_index.py

```diff
@@ -0,0 +1,151 @@
+
+    def _iter(self, from_timestamp: Optional[int] = None, from_tx: Optional[bytes] = None,
+              *, reverse: bool = False) -> Iterator[Tuple[int, bytes]]:
+        """ Iterate over all transactions of a token, by default from oldest to newest.
+        """
+        if from_timestamp is None and from_tx is not None:
+            raise ValueError('from_tx needs from_timestamp, but it is None')
+        it = self._db.iterkeys(self._cf)
+        if reverse:
+            it = reversed(it)
+            if from_timestamp is None:
+                self.log.debug('seek to last')
+                it.seek_to_last()
+            else:
+                # when reversed we increment the key by 1, which effectively goes to the end of a prefix
+                self.log.debug('seek to', tx=from_tx)
+                it.seek_for_prev(incr_key(self._to_key(from_timestamp, from_tx)))
+        else:
+            if from_timestamp is None:
+                self.log.debug('seek to first')
+                it.seek_to_first()
+            else:
+                self.log.debug('seek to', tx=from_tx)
+                it.seek(self._to_key(from_timestamp, from_tx))
+        first = True
+        for _, key in it:
+            timestamp, tx_hash = self._from_key(key)
+            self.log.debug('seek found', tx=tx_hash.hex())
+            # XXX: this is made such that we wouldn't skip the first element blindly, only if it is the given from_tx
+            if first:
+                first = False
+                # from_tx transaction is never yielded
+                if from_timestamp is not None and from_timestamp == timestamp and from_tx == tx_hash:
+                    continue
+            yield timestamp, tx_hash
+        self.log.debug('seek end')
+
+    def get_newest(self, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(reverse=True))
+        return collect_n(it, count)
+
+    def get_older(self, timestamp: int, hash_bytes: bytes, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(timestamp, hash_bytes, reverse=True))
+        return collect_n(it, count)
+
+    def get_newer(self, timestamp: int, hash_bytes: bytes, count: int) -> Tuple[List[bytes], bool]:
+        it = (x for _, x in self._iter(timestamp, hash_bytes))
+        return collect_n(it, count)
+
+    def get_hashes_and_next_idx(self, from_idx: RangeIdx, count: int) -> Tuple[List[bytes], Optional[RangeIdx]]:
+        assert count > 0, 'count<=0 not allowed'
+        timestamp, offset = from_idx
+        it = skip_n(self._iter(timestamp), offset)
+        hashes: List[bytes] = []
+        n = count
+        next_timestamp = None
+        next_offset = offset
+        while n > 0:
+            try:
+                next_timestamp, tx_hash = next(it)
+            except StopIteration:
+                return hashes, None
+            hashes.append(tx_hash)
+            if next_timestamp != timestamp:
+                next_offset = 0
```

Comment:
If we have more than one hash in the "new" next timestamp, shouldn't we return the correct offset instead of always returning 0? What's the expected behaviour here?

Reply:
The way offset works is on top of the timestamp query. For example, consider these timestamps+"hashes":

- 1 `aaaa`
- 1 `bbbb`
- 2 `cccc`
- 2 `dddd`
- 2 `eeee`
- 4 `ffff`

A timestamp+offset tuple of (1, 0) will start at `aaaa`, (1, 1) at `bbbb`, (2, 0) `cccc`, (2, 1) `dddd`, (2, 2) `eeee`, (4, 0) `ffff`, and so on.

This means that when iterating, every time we change the timestamp, the offset is reset to 0. And that's how I implemented this iterator. On the memory implementation we have access to the indexes and use that to calculate the offset, but on RocksDB there is no equivalent to a list index, so it was easier to just use this loop, since we have to iterate over the items anyway to collect the hashes.

Reply:
After our talk I got what you meant. Although I fixed this, I'm having some trouble making a test. I have a prototype for a tests but I can't get it to fail with the "wrong" implementation.